From 2a9051ad9ac6dc49fe79bcda1731a4de7ca1b127 Mon Sep 17 00:00:00 2001
From: Antonio Borneo <antonio.borneo@foss.st.com>
Date: Wed, 5 Feb 2025 15:32:46 +0100
Subject: [PATCH 1595/1695] plat-stm32mp1: make PWR IRQ independent from EXTI

The PWR IRQ is an interrupt controller that is hierarchical placed
between EXTI and GIC.
While still handle the current code that calls EXTI API, let the
driver PWR IRQ to work standalone relying to the hierarchical
calls.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git b5ff74d2114bc4820f5d116af3adedef72aa6053 plat-stm32mp1: make PWR IRQ independent from EXTI]
Signed-off-by: Antonio Borneo <antonio.borneo@foss.st.com>
Change-Id: I659a44a63d81d5521a0b0db45c1c57d679715959
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/434936
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr_irq.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr_irq.c b/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr_irq.c
index b2dd72c2b..0d49758fa 100644
--- a/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr_irq.c
+++ b/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr_irq.c
@@ -154,7 +154,7 @@ static void stm32mp1_pwr_itr_enable_nolock(size_t it)
 
 	VERBOSE_PWR("Pwr irq enable");
 
-	if (IS_ENABLED(CFG_STM32_EXTI)) {
+	if (IS_ENABLED(CFG_STM32_EXTI) && priv->exti) {
 		stm32_exti_unmask(priv->exti, PWR_EXTI_WKUP1 + it);
 		stm32_exti_enable_wake(priv->exti, PWR_EXTI_WKUP1 + it);
 	}
@@ -170,7 +170,7 @@ static void stm32mp1_pwr_itr_disable_nolock(size_t it)
 
 	io_clrbits32(priv->base + MPUWKUPENR, BIT(it));
 
-	if (IS_ENABLED(CFG_STM32_EXTI)) {
+	if (IS_ENABLED(CFG_STM32_EXTI) && priv->exti) {
 		stm32_exti_mask(priv->exti, PWR_EXTI_WKUP1 + it);
 		stm32_exti_disable_wake(priv->exti, PWR_EXTI_WKUP1 + it);
 	}
@@ -412,7 +412,7 @@ static TEE_Result stm32mp1_pwr_itr_dt_get(struct dt_pargs *args,
 		return res;
 	}
 
-	if (IS_ENABLED(CFG_STM32_EXTI))
+	if (IS_ENABLED(CFG_STM32_EXTI) && priv->exti)
 		stm32_exti_set_tz(priv->exti, PWR_EXTI_WKUP1 + itr_num);
 
 	itr_desc_p->chip = &stm32mp1_pwr_itr_chip;
@@ -446,7 +446,7 @@ static TEE_Result stm32mp1_pwr_irq_probe(const void *fdt, int node,
 	if (IS_ENABLED(CFG_STM32_EXTI)) {
 		res = stm32_exti_get_pdata(fdt, node, &priv->exti);
 		if (res)
-			goto err;
+			priv->exti = NULL;
 	}
 
 	res = interrupt_dt_get(fdt, node, &itr_chip, &itr_num);
-- 
2.47.2

